### PR TITLE
TensorBoardLogger sub_dir parameter for grouping logs

### PR DESCRIPTION
**Pull request by janhenriklambrechts**
## What does this PR do?

**TLDR:** Adds a `sub_dir` parameter for the `TensorBoardLogger` useful for grouping multiple runs in one folder as per #6019.

Although in #6019 I advocated for a general `sub_dir` arg for all logger instances to I eventually added it as a `TensorBoardLogger`-only parameter. Therefore, this PR only solves the "group multiple tensorboard logs into one"-aspect and not the "keep different logger-type logs in different folders"-aspect. It would still be cool to arrive at the solution described in #6019  - yet that will take a bigger overhaul of the logging module (f.e. adding version parameter to all loggers).

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/loggers/tensorboard.py [hunk 1]
```diff
@@ -57,6 +57,9 @@ class TensorBoardLogger(LightningLoggerBase):
             directory for existing versions, then automatically assigns the next available version.
             If it is a string then it is used as the run-specific subdirectory name,
             otherwise ``'version_${version}'`` is used.
+        sub_dir: Sub-directory to group TensorBoard logs. If a sub_dir argument is passed
+            then logs are saved in ``/save_dir/version/``. Defaults to ``None`` in which
```
Comment (justusschock) [inline pytorch_lightning/loggers/tensorboard.py:61]:
```suggestion
            then logs are saved in ``/save_dir/version/sub_dir/``. Defaults to ``None`` in which
```

At least that's what I assume, since otherwise this would be pointless :)

File: pytorch_lightning/loggers/tensorboard.py [hunk 2]
```diff
@@ -75,6 +78,7 @@ def __init__(
         save_dir: str,
         name: Optional[str] = "default",
         version: Optional[Union[int, str]] = None,
+        sub_dir: Optional[str] = None,
```
Comment (justusschock) [inline pytorch_lightning/loggers/tensorboard.py:81]: Not sure about this, but we may want to add this as the last argument here, since otherwise it may be breaking for people using positional arguments for init.

**Review by justusschock — COMMENTED** [review]
Two minor comments and one question:

Why can't we just add this to `save_dir`?

If you want to provide the argument you have to instantiate the logger manually anyways and then you can also alter the default directory given to the trainer.

File: pytorch_lightning/loggers/tensorboard.py (hunk 2 above)
Comment (janhenriklambrechts) [reply pytorch_lightning/loggers/tensorboard.py:81]: It really depends on what to prioritize, I prefer this order because it makes sense semantically - i.e. the order  corresponds to the actual logging path: `save_dir / name / version / sub_dir`. However, if making sure that nothing breaks is a top priority for PL then I will edit the PR.

File: pytorch_lightning/loggers/tensorboard.py (hunk 1 above)
Comment (janhenriklambrechts) [reply pytorch_lightning/loggers/tensorboard.py]: Ah my bad, thanks for spotting this!

**Comment by janhenriklambrechts** [comment]
@justusschock thank you for reviewing! 

First of all, let me confirm that you mean an approach in which you pass: `save_dir= [normal_path, sub_dir]`.
It is a valid question, but the reason why I settled on this approach is because it is a `TensorBoardLogger`-only argument. If we alter the behaviour of `save_dir` here it will not be consistent with other loggers `save_dir` because they will not have this functionality.

File: pytorch_lightning/loggers/tensorboard.py [hunk 3]
```diff
@@ -57,6 +57,9 @@ class TensorBoardLogger(LightningLoggerBase):
             directory for existing versions, then automatically assigns the next available version.
             If it is a string then it is used as the run-specific subdirectory name,
             otherwise ``'version_${version}'`` is used.
+        sub_dir: Sub-directory to group TensorBoard logs. If a sub_dir argument is passed
+            then logs are saved in ``/save_dir/version/sub_dir/``. Defaults to ``None`` in which
```
Comment (tchaton) [inline pytorch_lightning/loggers/tensorboard.py:61]: Wont it make more sense to save `/save_dir/sub_dir/version_{count}` ?

**Review by tchaton — APPROVED** [review]

File: pytorch_lightning/loggers/tensorboard.py [hunk 4]
```diff
@@ -114,13 +119,18 @@ def log_dir(self) -> str:
         """
         # create a pseudo standard path ala test-tube
         version = self.version if isinstance(self.version, str) else f"version_{self.version}"
-        log_dir = os.path.join(self.root_dir, version)
+        sub_dir = self.sub_dir if isinstance(self.sub_dir, str) else ""
+        log_dir = os.path.join(self.root_dir, version, sub_dir)
```
Comment (tchaton) [inline pytorch_lightning/loggers/tensorboard.py:123]: Should we check sub_dir doesn't contain `~` or `..`

Comment (justusschock) [reply pytorch_lightning/loggers/tensorboard.py:123]: I think you can just use the expand options for os in every case here.

File: pytorch_lightning/loggers/tensorboard.py (hunk 2 above)
Comment (justusschock) [reply pytorch_lightning/loggers/tensorboard.py:81]: Yes, we are very careful with ordering here.

File: pytorch_lightning/loggers/tensorboard.py (hunk 3 above)
Comment (janhenriklambrechts) [reply pytorch_lightning/loggers/tensorboard.py:61]: Not really, the purpose of `sub_dir` is to provide a folder similar to `checkpoints` that groups all logs **within a run** together.
See an example here: 
```
save_dir
├── version_0
│   ├── checkpoints
│   │   ├── checkpoint_0.ckpt
│   │   └── checkpoint_1.ckpt
│   └── sub_dir
│       ├── log_0.log
│       └── log_1.log
└── version_1
    ├── checkpoints
    │   ├── checkpoint_0.ckpt
    │   └── checkpoint_1.ckpt
    └── sub_dir
        ├── log_0.log
        └── log_1.log
```

File: pytorch_lightning/loggers/tensorboard.py (hunk 2 above)
Comment (janhenriklambrechts) [reply pytorch_lightning/loggers/tensorboard.py]: clear, position has been swapped in latest commit

File: pytorch_lightning/loggers/tensorboard.py (hunk 4 above)
Comment (janhenriklambrechts) [reply pytorch_lightning/loggers/tensorboard.py:123]: @justusschock I am not exactly following. You mean adding the `os.path.expand` option to every logger `log_dir` ?

File: pytorch_lightning/loggers/tensorboard.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/loggers/tensorboard.py:66]: Ok I get you. I thought it was to group several runs together.

**Comment by Borda** [comment]
@janhenriklambrechts how is it going here, ready to go? :rabbit:

**Comment by janhenriklambrechts** [comment]
@Borda for me, yes 😀

**Comment by janhenriklambrechts** [comment]
Seems like there are some tests failing now after merge. Will look into it tonight and fix :)

File: pytorch_lightning/loggers/tensorboard.py (hunk 4 above)
Comment (s-rog) [reply pytorch_lightning/loggers/tensorboard.py:128]:
```suggestion
        log_dir = os.path.join(self.root_dir, version, sub_dir)
        log_dir = os.path.expandvars(log_dir)
        log_dir = os.path.expanduser(log_dir)
```

I think this is what justus means, would also need tests for `~` and `$` in paths.

**Review by s-rog — CHANGES_REQUESTED** [review]
Just that one thing tchaton and justus mentioned above!

**Comment by janhenriklambrechts** [comment]
Actually, this one test that fails puzzles me:

```
def test_simple_profiler_log_dir(tmpdir):
       ... 
        expected = tmpdir / "lightning_logs" / "version_0"
        assert trainer.log_dir == expected

---- FAILED
> assert trainer.log_dir == expected
    AssertionError:
    +'.../version_0/'
    -'.../version_0'
```

Do not really see how this gets changed by this MR, or how to fix either the test or my feature accordingly.

@Borda could you maybe give me a pointer?

**Comment by s-rog** [comment]
@janhenriklambrechts I think that is failing from either a missing or extra `/` at the end of the path

**Comment by s-rog** [comment]
@janhenriklambrechts 
From the docs of `os.path.join`:
> ... will only end in a separator if the last part is empty

Example:
```
os.path.join("1", "2", "3")
'1/2/3'
os.path.join("1", "2", "")
'1/2/'
```
That's why it's outputting the extra separator in tests.

**Comment by janhenriklambrechts** [comment]
@s-rog thanks! Using this info to hotfix that test now :)

**Comment by s-rog** [comment]
Oh I already went ahead and fixed it! also, for the test can you actually pass in env vars?

**Comment by janhenriklambrechts** [comment]
@s-rog ah rip, just saw your message that you went ahead and fixed it - should I revert these commits?

**Comment by s-rog** [comment]
Yeah, I think it's best to not change the old test and have consistent separator usage regardless of sub dir usage.

lmk if you run into git issues and I'll see what I can do from my end.

**Comment by janhenriklambrechts** [comment]
@s-rog reverted - thanks for your help!
 This only leaves this:
> what do you exactly mean with env_vars - I did add a test for handling ~ and $ correctly. What is the desired behaviour for env vars?

**Review by s-rog — APPROVED** [review]
@janhenriklambrechts I went ahead and added it, should be good to merge once test pass!

**Comment by janhenriklambrechts** [comment]
@s-rog cool thanks - I watch and learn 😄

**Comment by s-rog** [comment]
@justusschock Your suggestion should be addressed if you wanna take another look
